src/core/search/ast_expr.h

```diff
@@ -4,21 +4,106 @@
 
 #pragma once
 
+#include <algorithm>
+#include <iostream>
+#include <memory>
 #include <ostream>
+#include <vector>
+
+#include "base/logging.h"
 
 namespace dfly {
 
 namespace search {
 
-class AstExpr {};
+// Describes a single node of the filter AST tree.
+class AstNode {
+ public:
+  virtual ~AstNode() = default;
+
+  // Check if this input is matched by the node.
+  virtual bool Check(std::string_view input) const = 0;
+
+  // Debug print node.
+  virtual std::string Debug() const = 0;
+};
+
+using AstNodePtr = std::unique_ptr<AstNode>;
+
+// Do we have to pollute all our parser file with moves?
+struct AstExpr {
+  AstExpr() : ptr{} {
+  }
+  AstExpr(AstNodePtr ptr) : ptr{move(ptr)} {
+  }
+  AstExpr(const AstExpr& other) {
+    ptr = move(other.ptr);
+  }
+  AstExpr& operator=(const AstExpr& other) {
+    ptr = move(other.ptr);
+    return *this;
+  }
+  operator AstNodePtr() {
+    return move(ptr);
+  }
+  const AstNode* operator->() const {
+    return ptr.get();
+  }
+  mutable AstNodePtr ptr;
+};
+
+template <typename T, typename... Ts> AstExpr MakeExpr(Ts... ts) {
+  AstNode* node = dynamic_cast<AstNode*>(new T(std::forward<Ts>(ts)...));
+  CHECK(node);
```

Comment:
This can be a compile time check, right?
Maybe do something like `static_assert(std::is_base_of_v<AstNode, T>, "...");`?
And if you do that, you can skip the dynamic cast, and even make this a one liner:

```
return std::make_unique<T>(std::forward<Ts>(ts)...);
```

I could very well be missing something, sorry if I am!